video/out/opengl/hwdec_cuda.c

```diff
@@ -0,0 +1,287 @@
+/*
+ * Copyright (c) 2016 Philip Langdale <philipl@overt.org>
+ *
+ * This file is part of mpv.
+ *
+ * mpv is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * mpv is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with mpv.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/*
+ * This hwdec implements an optimized output path using CUDA->OpenGL
+ * interop for frame data that is stored in CUDA device memory.
+ * Although it is not explicit in the code here, the only practical way
+ * to get data in this form is from the 'cuvid' decoder (aka NvDecode).
+ *
+ * For now, cuvid/NvDecode will always return images in NV12 format, even
+ * when decoding 10bit streams (there is some hardware dithering going on).
+ */
+
+#include <libavutil/hwcontext.h>
+#include <libavutil/hwcontext_cuda.h>
+
+#include "hwdec.h"
+#include "video.h"
+
+#include <cudaGL.h>
+
+struct priv {
+    struct mp_hwdec_ctx hwctx;
+    struct mp_image_params image_params;
+    GLuint gl_textures[2];
+    GLuint gl_pbos[2];
+    bool mapped;
+
+    CUcontext cuda_ctx;
+};
+
+static int check_cu(struct gl_hwdec *hw, CUresult err, const char *func)
+{
+    const char *err_name;
+    const char *err_string;
+
+    MP_TRACE(hw, "Calling %s\n", func);
+
+    if (err == CUDA_SUCCESS)
+        return 0;
+
+    cuGetErrorName(err, &err_name);
+    cuGetErrorString(err, &err_string);
+
+    MP_ERR(hw, "%s failed", func);
+    if (err_name && err_string)
+        MP_ERR(hw, " -> %s: %s", err_name, err_string);
+    MP_ERR(hw, "\n");
+
+    return -1;
+}
+
+#define CHECK_CU(x) check_cu(hw, (x), #x)
+
+static int cuda_create(struct gl_hwdec *hw)
+{
+    CUdevice device;
+    CUcontext cuda_ctx = NULL;
+    CUcontext dummy;
+    int ret = 0, eret = 0;
+
+    struct priv *p = talloc_zero(hw, struct priv);
+    hw->priv = p;
+
+    ret = CHECK_CU(cuInit(0));
+    if (ret < 0) {
+        goto error;
+    }
+
+    ///TODO: Make device index configurable
+    ret = CHECK_CU(cuDeviceGet(&device, 0));
+    if (ret < 0) {
+        goto error;
+    }
```

Comment:
Before I forgot it: if the body of an if statement is only one physical line, the `{`/`}` can be left away.